model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java

```diff
@@ -0,0 +1,293 @@
+
+    /**
+     * Get the maximum lifespan, which this userSession can remain in the infinispan cache.
+     * Returned value will be used as "lifespan" when calling put/replace operation in the infinispan cache for this entity
+     *
+     * @param realm
+     * @param userSessionEntity
+     * @return
+     */
+    public static long getUserSessionLifespanMs(RealmModel realm, UserSessionEntity userSessionEntity) {
+        int timeSinceSessionStart = Time.currentTime() - userSessionEntity.getStarted();
+
+        int sessionMaxLifespan = Math.max(realm.getSsoSessionMaxLifespan(), MINIMAL_EXPIRATION_SEC);
+        if (userSessionEntity.isRememberMe()) {
+            sessionMaxLifespan = Math.max(realm.getSsoSessionMaxLifespanRememberMe(), sessionMaxLifespan);
+        }
+
+        long timeToExpire = sessionMaxLifespan - timeSinceSessionStart;
+
+        // Indication that entry should be expired
+        if (timeToExpire <=0) {
+            return ENTRY_EXPIRED_FLAG;
+        }
+
+        return Time.toMillis(timeToExpire);
+    }
+
+
+    /**
+     * Get the maximum idle time for this userSession.
+     * Returned value will be used when as "maxIdleTime" when calling put/replace operation in the infinispan cache for this entity
+     *
+     * @param realm
+     * @param userSessionEntity
+     * @return
+     */
+    public static long getUserSessionMaxIdleMs(RealmModel realm, UserSessionEntity userSessionEntity) {
+        int timeSinceLastRefresh = Time.currentTime() - userSessionEntity.getLastSessionRefresh();
+
+        int sessionIdleMs = Math.max(realm.getSsoSessionIdleTimeout(), MINIMAL_EXPIRATION_SEC);
+        if (userSessionEntity.isRememberMe()) {
+            sessionIdleMs = Math.max(realm.getSsoSessionIdleTimeoutRememberMe(), sessionIdleMs);
+        }
+
+        long maxIdleTime = sessionIdleMs - timeSinceLastRefresh + SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+
+        // Indication that entry should be expired
+        if (maxIdleTime <=0) {
+            return ENTRY_EXPIRED_FLAG;
+        }
+
+        return Time.toMillis(maxIdleTime);
+    }
+
+
+    /**
+     * Get the maximum lifespan, which this clientSession can remain in the infinispan cache.
+     * Returned value will be used as "lifespan" when calling put/replace operation in the infinispan cache for this entity
+     *
+     * @param realm
+     * @param clientSessionEntity
+     * @return
+     */
+    public static long getClientSessionLifespanMs(RealmModel realm, AuthenticatedClientSessionEntity clientSessionEntity) {
+        int timeSinceTimestampUpdate = Time.currentTime() - clientSessionEntity.getTimestamp();
+
+        int sessionMaxLifespan = Math.max(realm.getSsoSessionMaxLifespan(), realm.getSsoSessionMaxLifespanRememberMe());
+
+        // clientSession max lifespan has preference if set
+        if (realm.getClientSessionMaxLifespan() > 0) {
+            sessionMaxLifespan = realm.getClientSessionMaxLifespan();
+        }
```

Comment:
@mposolda Don't you need to consider these value (ClientSessionMaxLifespan, ClientSessionIdleTimeout, ClientOfflineSessionMaxLifespan, ClientOfflineSessionIdleTimeout) are overridden per client?

Reply:
@y-tabata Thanks for pointing this. I've created https://issues.redhat.com/browse/KEYCLOAK-17111 for this. I see that before this PR, the InfinispanUserSessionProvider.removeExpired(realm) also ignored client-overriden timeouts, so this bug is not a regression of this PR. However it works fixing.